### ISSUE TITLE
type flow analysis false positive with change in function call

### ISSUE THREAD

**Issue by JounQin** [microsoft/TypeScript#45443]
# Bug Report

### 🔎 Search Terms

flow analysis

### 🕗 Version & Regression Information

- This is a crash
- This changed between versions ______ and _______
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________
- I was unable to test this on prior versions because _______

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.4.0-beta#code/KYDwDg9gTgLgBAYwgOwM7xsdcC8cAUAlLgHxwDeAUHHADbDwBuAhrbnAGaurCXWIpsHZOyKkK-GizZ4YUAK68aAXz41hRNXACWHAtOLk4AemNwABtPM7UcdNtptzXWj2sALYFGAAaCwCMICHpmZGtQMGAETAATSQE0YOAAOloIAHN8AHIswn5VVSA)

### 💻 Code

```ts
export const test = () => {
  let val = false

  const fn = () => {
    val = true
  }

  fn()

  if (val) { // `val` is still `false` here, `boolean` expected
    console.log('')
  }
}
```

### 🙁 Actual behavior

`val` is still `false`

### 🙂 Expected behavior

`val` should be `boolean`

---

related https://github.com/typescript-eslint/typescript-eslint/issues/3732

**Comment by MartinJohns** [comment]
And another duplicate of #9998.
